cpplint.py

```diff
@@ -979,26 +1006,38 @@ def ParseNolintSuppressions(filename, raw_line, linenum, error):
     linenum: int, the number of the current line.
     error: function, an error handler.
   """
-  matched = Search(r'\bNOLINT(NEXTLINE)?\b(\([^)]+\))?', raw_line)
+  matched = Search(r'\bNOLINT(NEXTLINE|BEGIN|END)?\b(\([^)]+\))?', raw_line)
   if matched:
-    if matched.group(1):
-      suppressed_line = linenum + 1
+    no_lint_type = matched.group(1)
+    if no_lint_type == 'NEXTLINE':
+      process_category = lambda c: _error_suppressions.setdefault(c, set()).add(linenum + 1)
+    elif no_lint_type == 'BEGIN':
+      process_category = lambda c: _block_error_suppressions.update((c,))
+    elif no_lint_type == 'END':
+      def process_category(category):
+        if _block_error_suppressions[category] <= 0:
+          error(filename, linenum, 'readability/nolint', 5,
+                'NOLINT category not suppressed in current block: %s' % (category or 'ALL'))
+        else:
+          _block_error_suppressions.subtract((category,))
     else:
-      suppressed_line = linenum
+      process_category = lambda c: _error_suppressions.setdefault(c, set()).add(linenum)
     category = matched.group(2)
     if category in (None, '(*)'):  # => "suppress all"
-      _error_suppressions.setdefault(None, set()).add(suppressed_line)
+      process_category(None)
     else:
       if category.startswith('(') and category.endswith(')'):
-        category = category[1:-1]
-        if category in _ERROR_CATEGORIES:
-          _error_suppressions.setdefault(category, set()).add(suppressed_line)
-        elif any(c for c in _OTHER_NOLINT_CATEGORY_PREFIXES if category.startswith(c)):
-          # Ignore any categories from other tools.
-          pass
-        elif category not in _LEGACY_ERROR_CATEGORIES:
-          error(filename, linenum, 'readability/nolint', 5,
-                'Unknown NOLINT error category: %s' % category)
+        categories = category[1:-1].split(',')
```

Comment:
Am I right that this also extends the ability of `NOLINT(...)`? If so, that should also be tested in unit tests, and documented.

Reply:
Possibly this handling of multiple categories on the same line could be extracted as a separate commit, or even separate MR. But I will accept either.

Reply:
I am thinking about handling duplicates here, should that be handled? Same as ignoring things that are already ignored?

Reply:
I will make a separate PR for adding multiple categories to one NOLINT line and then apply this PR on top of that one. We can focus on that piece first